File: Phase1/ProgrammingBasics/Cpp/Variables_And_Data_Structures.cpp
// Summary of Variables and Data Structures in C++ (for Robotics Programming)

#include <iostream>
#include <vector> // Dynamic array
#include <array>  // Fixed-size array
#include <string> // Text data
using namespace std;

int main() {
    // --- VARIABLES ---

    // Integer (robot ID, counter, etc.)
    int robotID = 1;

    // Float (position coordinates, speeds)
    float robotSpeed = 3.5f; // "f" suffix for float literals

    // Double (more precision for calculations)
    double distanceToGoal = 15.6789;

    // Boolean (status flags)
    bool isOperational = true;

    // Character (single character)
    char direction = 'N'; // N for North

    // String (useful for names, messages)
    string robotName = "Atlas";

    // Print variables
    cout << "Robot " << robotName << " [ID: " << robotID << "] is heading "
         << direction << " at speed " << robotSpeed << " m/s." << endl;

    // --- DATA STRUCTURES ---

    // 1. Arrays (fixed-size collection of elements)
    int sensorValues[5] = {100, 95, 90, 85, 80};
    cout << "First sensor reading: " << sensorValues[0] << endl;

    // 2. std::array (safer fixed-size array)
    array<int, 3> wheelSpeeds = {10, 20, 30};
    cout << "Second wheel speed: " << wheelSpeeds[1] << endl;

    // 3. std::vector (dynamic array)
    vector<float> lidarReadings;
    lidarReadings.push_back(1.2);
    lidarReadings.push_back(2.5);
    lidarReadings.push_back(0.9);

    cout << "LIDAR reading 1: " << lidarReadings[0] << endl;

    // 4. Structs (group related variables together)
    struct Position {
        double x;
        double y;
        double theta; // orientation in radians
    };

    Position robotPosition = {0.0, 0.0, 0.0};
    robotPosition.x = 5.2;
    robotPosition.y = 3.8;
    robotPosition.theta = 1.57; // ~90 degrees

    cout << "Robot Position -> x: " << robotPosition.x
         << ", y: " << robotPosition.y
         << ", theta: " << robotPosition.theta << " radians" << endl;

    // --- BONUS: CONSTANTS ---
    // Useful for things that shouldn't change (e.g., robot parameters)
    const double WHEEL_RADIUS = 0.15; // meters
    // WHEEL_RADIUS = 0.2; // Error! Cannot modify const variables

    return 0;
}

File: Phase1/ProgrammingBasics/Python/Variables_And_Data_Structures.py
# Purpose: Summary of Variables, Data Types, and Data Structures for Robotics Programming in Python

# -----------------------------
# 1. VARIABLES
# -----------------------------

# Variables store information the robot needs (position, speed, sensor data, etc.)
robot_name = "ExplorerBot"       # String
battery_level = 87.5              # Float
is_moving = True                  # Boolean
wheel_count = 4                   # Integer

# -----------------------------
# 2. BASIC DATA TYPES
# -----------------------------

# Integers: whole numbers (used for counting, discrete steps)
steps_moved = 120

# Floats: numbers with decimals (used for precise measurements)
motor_speed = 2.5   # meters per second

# Strings: text (used for IDs, labels)
sensor_type = "Infrared"

# Booleans: True/False (used for checking robot states)
obstacle_detected = False

# -----------------------------
# 3. COLLECTION DATA TYPES (important for grouping robot data)
# -----------------------------

# Lists: Ordered collections (good for multiple sensors, motor speeds)
sensor_readings = [23.4, 22.9, 24.1]  # readings from temperature sensors

# Access elements
first_reading = sensor_readings[0]  # 23.4

# Modify elements
sensor_readings[1] = 23.0

# Append new readings
sensor_readings.append(24.5)

# Iterate through a list
for reading in sensor_readings:
    print("Sensor reading:", reading)

# -----------------------------
# Tuples: Immutable collections (good for fixed coordinates)
# -----------------------------

position = (10, 5)  # (x, y) position of robot

# Access tuple elements
x_coord = position[0]
y_coord = position[1]

# Tuples can't be changed, which is good for constant positions

# -----------------------------
# Dictionaries: Key-value pairs (good for labeling sensor data)
# -----------------------------

sensor_data = {
    "front_sensor": 1.5,  # meters
    "rear_sensor": 2.0,
    "left_sensor": 0.8
}

# Access dictionary values
front_distance = sensor_data["front_sensor"]

# Add new data
sensor_data["right_sensor"] = 1.2

# Iterate through dictionary
for sensor, distance in sensor_data.items():
    print(sensor, "distance:", distance)

# -----------------------------
# Sets: Unordered, unique items (less common, but useful for tracking unique events)
# -----------------------------

visited_zones = {"A", "B", "C"}

# Add a new zone
visited_zones.add("D")

# Check if a zone has been visited
if "B" in visited_zones:
    print("Zone B has been visited.")

# -----------------------------
# 4. TYPE CONVERSION (important for sensor data)
# -----------------------------

# Example: sensor gives string output, but we need a float
sensor_output = "23.7"
sensor_value = float(sensor_output)

# Example: motor ID integer to string
motor_id = 5
motor_label = "Motor_" + str(motor_id)

# -----------------------------
# 5. CONSTANTS (optional best practice)
# -----------------------------

# Constants: variables you shouldn't change (Python doesn't enforce this, but ALL_CAPS by convention)
MAX_SPEED = 5.0  # meters per second

# Example usage
if motor_speed > MAX_SPEED:
    print("Warning: Exceeding maximum speed!")

# -----------------------------
# End of script
# -----------------------------

# This covers all fundamental variable/data concepts you'll use when programming a robot!
# Practice Tip: Try to simulate a small robot control system using these types!
